keps/sig-cli/0031-datadrivencommands.md

```diff
@@ -0,0 +1,412 @@
+
+	// Output is a go-template used write the command output.  It may reference values specified as flags using
+	// {{index .Flags.Strings "flag-name"}}, {{index .Flags.Ints "flag-name"}}, {{index .Flags.Bools "flag-name"}},
+	// {{index .Flags.Floats "flag-name"}}.
+	//
+	// It may also reference values from the responses that were saved using saveResponseValues
+	// - {{index .Responses.Strings "response-value-name"}}.
+	//
+	// Example:
+	// 		deployment.apps/{{index .Responses.Strings "responsename"}} created
+	//
+	// +optional
+	Output string `json:"output,omitempty"`
+}
+
+type ResourceOperation string
+
+const (
+	CREATE_RESOURCE ResourceOperation = "Create"
+	UPDATE_RESOURCE                   = "Update"
+	DELETE_RESOURCE                   = "Delete"
+	GET_RESOURCE                      = "Get"
+	PATCH_RESOURCE                    = "Patch"
+)
+
+type ResourceRequest struct {
+	// Group is the API group of the request endpoint
+	//
+	// Example: apps
+	Group string `json:"group"`
+
+	// Version is the API version of the request endpoint
+	//
+	// Example: v1
+	Version string `json:"version"`
+
+	// Resource is the API resource of the request endpoint
+	//
+	// Example: deployments
+	Resource string `json:"resource"`
+
+	// Operation is the type of operation to perform for the request.  One of: Create, Update, Delete, Get, Patch
+	Operation ResourceOperation `json:"operation"`
+
+	// BodyTemplate is a go-template for the request Body.  It may reference values specified as flags using
+	// {{index .Flags.Strings "flag-name"}}, {{index .Flags.Ints "flag-name"}}, {{index .Flags.Bools "flag-name"}},
+	// {{index .Flags.Floats "flag-name"}}
+	//
+	// Example:
+	//      apiVersion: apps/v1
+	//      kind: Deployment
+	//      metadata:
+	//        name: {{index .Flags.Strings "name"}}
+	//        namespace: {{index .Flags.Strings "namespace"}}
+	//        labels:
+	//          app: nginx
+	//      spec:
+	//        replicas: {{index .Flags.Ints "replicas"}}
+	//        selector:
+	//          matchLabels:
+	//            app: {{index .Flags.Strings "name"}}
+	//        template:
+	//          metadata:
+	//            labels:
+	//              app: {{index .Flags.Strings "name"}}
+	//          spec:
+	//            containers:
+	//            - name: {{index .Flags.Strings "name"}}
+	//              image: {{index .Flags.Strings "image"}}
+	//
+	// +optional
+	BodyTemplate string `json:"bodyTemplate,omitempty"`
+
+	// SaveResponseValues are values read from the response and saved in {{index .Responses.Strings "flag-name"}}.
+	// They may be used in the ResourceCommand.Output go-template.
+	//
+	// Example:
+	//		- name: responsename
+	//        jsonPath: "{.metadata.name}"
+	//
+	// +optional
+	SaveResponseValues []ResponseValue `json:"saveResponseValues,omitempty"`
+}
+
+// Flag defines a cli flag that should be registered and available in request / output templates
+type Flag struct {
+	Type FlagType `json:"type"`
```

Comment:
We should align this closer to the structure of Table and be as consistent as possible.  Would prefer at minimum defining Format.

Reply:
Do you have a link to the format you are referring to?

Reply:
To clarify, this struct is for populating the client side command flags (e.g. Cobra).

Reply:
Added comments about this to clarify.

Reply:
Yeah, I meant that between flags, openapi, and the table object that server side get returns we should be consistent in how we describe and use constants.  Don’t have to be identical, but don’t drift for the sake of drift.

Reply:
SG.  Will add a TODO for this.